mkosi/__init__.py

```diff
@@ -3262,39 +3243,15 @@ def build_stuff(uid: int, gid: int, config: MkosiConfig) -> None:
     # while we are working on it.
     with flock(workspace_dir), workspace:
         # If caching is requested, then make sure we have cache trees around we can make use of
-        if need_cache_trees(state):
-
-            # There is no point generating a pre-dev cache image if no build script is provided
-            if config.build_script:
-                with complete_step("Running first (development) stage to generate cached copy…"):
-                    # Generate the cache version of the build image, and store it as "cache-pre-dev"
-                    state = dataclasses.replace(state, do_run_build_script=True, for_cache=True)
-                    build_image(state)
-                    save_cache(state)
-                    remove_artifacts(state)
-
-            with complete_step("Running second (final) stage to generate cached copy…"):
-                # Generate the cache version of the build image, and store it as "cache-pre-inst"
-                state = dataclasses.replace(state, do_run_build_script=False, for_cache=True)
+        if need_cache_tree(state):
+            with complete_step("Building cache image"):
+                state = dataclasses.replace(state, for_cache=True)
```

Comment:
I think having a different name here makes this easier to reason about, since, maybe for lack of a better expressions, it's "more immutable". `dataclass.replace` will not mutate `state`, but will make a new instance that is with the changes applied, so the logic is
$$s \to s'=sf(s) \to s=f^{-1}(s')=f^{-1}(f(s))$$
but under the hood it's
$$s \to s'=f(s) \to s''=g(s')\equiv s$$
where we have to separately make sure that $g$ is $f^{-1}$ so that $s \equiv s''$. By making both separate and reusing `state` below (and not having to do the second `dataclass.replace`, i.e. $g$) it's easy to follow that `state` ($s$) is the same before and after the build script step and "`state_for_cache`" (or however one would call it) is this other thing we're not going to mutate again or use for anything but this step.

Reply:
Let's address this is a separate PR, I don't wanna wait for CI again